File: mj_formatter/core/policy/conflict_detector.py
from __future__ import annotations

from ..types import ConflictDetectorConfig, Edit, Violation


class PolicyConflictDetector:
    def __init__(self, config: ConflictDetectorConfig | None = None) -> None:
        self._config = config or ConflictDetectorConfig()
        self._line_history: dict[int, list[Edit]] = {}
        self._reported_reverts: set[tuple[int, str, str]] = set()
        self._reported_touches: set[tuple[int, tuple[str, ...]]] = set()

    def observe(self, policy_name: str, edits: list[Edit]) -> list[Violation]:
        if not self._config.enabled or not edits:
            return []
        violations: list[Violation] = []
        for edit in edits:
            history = self._line_history.setdefault(int(edit.line), [])

            if history:
                prev = history[-1]
                if prev.policy != policy_name and edit.after == prev.before and edit.before == prev.after:
                    key = (int(edit.line), prev.policy, policy_name)
                    if key not in self._reported_reverts:
                        self._reported_reverts.add(key)
                        violations.append(
                            Violation(
                                policy="policy_conflict_detector",
                                message=(
                                    f"Line {edit.line}: policy '{policy_name}' appears to revert "
                                    f"policy '{prev.policy}'"
                                ),
                                line=edit.line,
                                column=1,
                            )
                        )

                touched = sorted({item.policy for item in history} | {policy_name})
                if len(touched) >= self._config.touch_threshold:
                    touch_key = (int(edit.line), tuple(touched))
                    if touch_key not in self._reported_touches:
                        self._reported_touches.add(touch_key)
                        violations.append(
                            Violation(
                                policy="policy_conflict_detector",
                                message=(
                                    f"Line {edit.line}: touched by multiple policies ({', '.join(touched)}), "
                                    "verify style rule compatibility"
                                ),
                                line=edit.line,
                                column=1,
                            )
                        )

            history.append(edit)
        return violations

File: src/engine/conflict_detector.rs
use rustc_hash::{FxHashMap, FxHashSet};

use crate::model::edit::Edit;
use crate::model::violation::Violation;

pub struct PolicyConflictDetector {
    enabled: bool,
    touch_threshold: usize,
    line_history: FxHashMap<usize, Vec<Edit>>,
    reported_reverts: FxHashSet<(usize, String, String)>,
    reported_touches: FxHashSet<(usize, String)>,
}

impl PolicyConflictDetector {
    pub fn new(enabled: bool, touch_threshold: usize) -> Self {
        Self {
            enabled,
            touch_threshold: touch_threshold.max(2),
            line_history: FxHashMap::default(),
            reported_reverts: FxHashSet::default(),
            reported_touches: FxHashSet::default(),
        }
    }

    pub fn observe(&mut self, policy_name: &str, edits: &[Edit]) -> Vec<Violation> {
        if !self.enabled || edits.is_empty() {
            return Vec::new();
        }

        let mut violations = Vec::<Violation>::new();
        for edit in edits {
            if edit.line == 0 {
                continue;
            }

            let history = self.line_history.entry(edit.line).or_default();
            if let Some(previous) = history.last() {
                if previous.policy != policy_name
                    && edit.after == previous.before
                    && edit.before == previous.after
                {
                    let key = (
                        edit.line,
                        previous.policy.to_string(),
                        policy_name.to_string(),
                    );
                    if self.reported_reverts.insert(key.clone()) {
                        violations.push(Violation {
                            policy: "policy_conflict_detector".into(),
                            message: format!(
                                "Line {}: policy '{}' appears to revert policy '{}'",
                                edit.line, policy_name, previous.policy
                            ),
                            line: edit.line,
                            column: Some(1),
                        });
                    }
                }

                let mut touched = history
                    .iter()
                    .map(|item| item.policy.as_str())
                    .collect::<Vec<_>>();
                touched.push(policy_name);
                touched.sort_unstable();
                touched.dedup();

                if touched.len() >= self.touch_threshold {
                    let touched_key = touched.join("|");
                    let key = (edit.line, touched_key.clone());
                    if self.reported_touches.insert(key) {
                        violations.push(Violation {
                            policy: "policy_conflict_detector".into(),
                            message: format!(
                                "Line {}: touched by multiple policies ({}), verify style rule compatibility",
                                edit.line,
                                touched.join(", ")
                            ),
                            line: edit.line,
                            column: Some(1),
                        });
                    }
                }
            }

            history.push(edit.clone());
        }

        violations
    }
}

#[cfg(test)]
mod tests {
    use crate::model::edit::Edit;

    use super::PolicyConflictDetector;

    #[test]
    fn reports_revert_conflict() {
        let mut detector = PolicyConflictDetector::new(true, 3);
        let first = Edit {
            policy: "p1".into(),
            line: 10,
            before: "A".to_string(),
            after: "B".to_string(),
        };
        let second = Edit {
            policy: "p2".into(),
            line: 10,
            before: "B".to_string(),
            after: "A".to_string(),
        };
        assert!(detector.observe("p1", &[first]).is_empty());
        let violations = detector.observe("p2", &[second]);
        assert_eq!(violations.len(), 1);
        assert!(violations[0].message.contains("appears to revert"));
    }
}
